source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc

```diff
@@ -611,6 +611,159 @@ void UdpProxyFilter::ActiveSession::writeDownstream(Network::UdpRecvData& recv_d
   }
 }
 
+void HttpUpstreamImpl::encodeData(Buffer::Instance& data) {
+  if (!request_encoder_) {
+    return;
+  }
+
+  request_encoder_->encodeData(data, false);
+}
+
+void HttpUpstreamImpl::setRequestEncoder(Http::RequestEncoder& request_encoder, bool is_ssl) {
+  request_encoder_ = &request_encoder;
+  request_encoder_->getStream().addCallbacks(*this);
+
+  const std::string& scheme =
+      is_ssl ? Http::Headers::get().SchemeValues.Https : Http::Headers::get().SchemeValues.Http;
+
+  std::string host = tunnel_config_.proxyHost(downstream_info_);
+  const auto* dynamic_port =
+      downstream_info_.filterState()->getDataReadOnly<StreamInfo::UInt32Accessor>(
+          "udp.connect.proxy_port");
+  if (dynamic_port != nullptr && dynamic_port->value() > 0 && dynamic_port->value() <= 65535) {
+    absl::StrAppend(&host, ":", std::to_string(dynamic_port->value()));
+  } else if (tunnel_config_.proxyPort().has_value()) {
+    absl::StrAppend(&host, ":", std::to_string(tunnel_config_.proxyPort().value()));
+  }
+
+  auto headers = Http::RequestHeaderMapImpl::create();
+  headers->addReferenceKey(Http::Headers::get().Scheme, scheme);
+  headers->addReferenceKey(Http::Headers::get().Host, host);
+
+  if (tunnel_config_.usePost()) {
```

Comment:
In case ``usePost()`` is used, we should consider this as not complying with connect-udp protocol, rather than just another method of tunneling. I will be using this to tunnel UDP over gRPC streams, where the encapsulation will be done in Protobuf  messages, and the headers has POST method, as per the gRPC scehme.

Reply:
Customization knobs are a burden for maintenance. If you're not planning on using `usePost` with connect-udp then I wouldn't add it.

Reply:
> If you're not planning on using usePost with connect-udp then I wouldn't add it.

It just has nothing to do with connect-udp - when using ``usePost``, the upstream stream is pivoted from connect-udp to simple POST request. In other words, by default - the ``udp_proxy`` would tunnel with respect to the connect-udp protocol, unless ``usePost`` is specified, then - it's comes to a simple HTTP POST request, which can be customized to support other use cases. This will be clearly specified in the docs and API that will be added in the next PR.

This kind of option has a value in my perspective and will be used in production. Since I don't see another way to add this and given that this isn't much of overhead, I'm inclined to keep it.

Reply:
This encourages uses of HTTP that go against the HTTP specs, so I think it should be avoided. But I won't stand in the way

Reply:
Right, I updated the description. The use case I am willing to support here and which will be used in production (we seek migrating to Envoy for this use case, therefore this flag is neeed) is tunneling UDP over gRPC streams, where the encapsulation would be done by a Protobuf message rather than an HTTP capsule. So in that case ``usePost`` would pivot the HTTP stream from ``connect-udp`` upgrade to POST request.
It can co-exist with the ``connect-udp`` upgrade, which will be the default behavior, and practically, the overhead of this flag is merely few lines of source, I will be super happy to keep it as is, because it doesn't diminish from the ``connect-udp`` upgrade, and serves a very much needed use case (for my org at least). Thanks

Reply:
Ah! Thanks for the context. If there's a use case in production, that seems compelling. (It still makes me sad to use HTTP this way, but ... welcome to the internet :>).